### PR TITLE
fix(gitlab): fallback to MR note for lockfiles and line_code errors

### PR DESCRIPTION
**Pull request by aqua-princewill**
# Description

[JIRA-132403](https://scalock.atlassian.net/browse/SLK-132403)

- Fix GitLab MR comment failures on packages.lock.json and other dependency lockfiles by falling back to a general MR note when inline placement is not possible.
- Replace fragile substring lockfile matching with basename-based detection for common npm, NuGet, Go, Rust, Python, and other lock/manifest files.
- Also fall back when GitLab returns a line_code validation error, which indicates the target line is not on the MR diff.
- Update the fallback message to reflect inline placement failure (not only large diffs).
- Add unit tests for fallback decision logic.

**Problem**
When Aqua posts vulnerability comments on GitLab MRs for .NET repos, inline comments on packages.lock.json fail with:

> line_code can't be blank / must be a valid line code

The existing fallback to a general MR note only applied to package.json and yarn.lock (via substring match). NuGet's packages.lock.json did not match, so every finding retried and failed. With many lockfiles this triggered hundreds of 400s and GitLab's elevated 400 error rates circuit breaker — no comments were posted.

## Type of change

- [ ] Chore (README.md update / pipeline update / etc)
- [ ] Bug fix **(non-breaking change which fixes an issue)**
- [ ] Vulnerability fix **(non-breaking change without code changes)**
- [ ] New feature **(non-breaking change which adds functionality)**
- [ ] Breaking change **(_(major version increased)_ - no backward compatibility)**

## Testing

- [x] Added / Ran automated tests **(unit / integration / e2e)**
- [ ] Manual tests - **local environment**
- [ ] Manual tests - **dev/staging environment**

## Release

- [ ] Documentation update required? **_(Notified PM on required changes)_**
- [ ] **Requires UI / CICD client / scanner alignment?**
- [ ] **Requires infrastructure alignment?**
- [ ] **Requires migration?**
- [ ] **Has Datadog monitoring?**
- [ ] **No new HIGH / CRITICAL vulnerabilities? _(FIPS ready / Updated KNOW_VULNERABILITIES.md)_**

## Screenshots / Video

### CODE REVIEW

**Comment by mjshastha** [comment]
@bugbot run

File: pkg/commenter/gitlab/gitlab.go [hunk 1]
```diff
@@ -49,7 +50,39 @@ type Gitlab struct {
...
+	"gemfile.lock":       {},
+	"podfile.lock":       {},
+	"pubspec.lock":       {},
+	"mix.lock":           {},
+	"shard.lock":         {},
+	"carton.lock":        {},
+	"luarocks.lock":      {},
+}
+
+func shouldUseGeneralPrComment(file, gitlabError string) bool {
```
Comment (mjshastha) [inline pkg/commenter/gitlab/gitlab.go]: Can we narrow this check instead of treating any error containing line_code as a reason to fall back to an MR-level comment?

line_code can also be returned for invalid lines/paths on regular source files. With the current logic, a bad inline position on something like src/main.go can result in a general MR note and WriteLineComment returning nil, so the caller believes the requested inline comment succeeded.

I think we should either match the specific GitLab validation errors we're targeting or limit this fallback to the intended lockfile cases.

File: pkg/commenter/gitlab/gitlab_test.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
+package gitlab
+
+import "testing"
+
+func TestShouldUseGeneralPrComment(t *testing.T) {
```
Comment (mjshastha) [inline pkg/commenter/gitlab/gitlab_test.go]: Can we add a case where a normal source file gets a line_code validation error?

File: pkg/commenter/gitlab/gitlab_test.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+		want        bool
+	}{
+		{
+			name: "nuget lock file",
+			file: "ELMS.Api.Tests/packages.lock.json",
+			want: true,
+		},
+		{
+			name: "npm lock file via package.json basename match",
+			file: "e2e/package-lock.json",
```
Comment (mjshastha) [inline pkg/commenter/gitlab/gitlab_test.go]: add a test for an empty/whitespace-only comment and verify that it returns nil without making any GitLab API calls. This directly covers the null/empty comment requirement from SLK-132403.

File: pkg/commenter/gitlab/gitlab_test.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+			file: "src/main.go",
+			gitlabError: `{"message":"400 Bad request - Note {:line_code=>[\"can't be blank\", \"must be a valid line code\"]}"}`,
+			want: true,
+		},
+		{
+			name: "unrelated error on regular source file",
+			file: "src/main.go",
+			gitlabError: `{"message":"401 Unauthorized"}`,
+			want: false,
+		},
```
Comment (mjshastha) [inline pkg/commenter/gitlab/gitlab_test.go]: The helper tests cover the new matching logic, but I'd like to see at least one httptest covering the actual flow:

getLatestVersion → inline attempt → inline failure → MR-note fallback

This would give us confidence that the HTTP responses, fallback behavior, and returned error/success status all work together.

File: pkg/commenter/gitlab/gitlab.go [hunk 5]
```diff
@@ -126,9 +159,8 @@ func (c *Gitlab) WriteLineComment(file, comment string, line int) error {
 			return err
 		}
 		if resp.StatusCode != http.StatusCreated {
-			if lo.ContainsBy(lockFiles, func(lf string) bool {
-				return strings.Contains(string(file), lf)
-			}) {
+			b, _ := io.ReadAll(resp.Body)
```
Comment (mjshastha) [inline pkg/commenter/gitlab/gitlab.go]: I don't see the response bodies being closed on these client.Do() calls. This code runs once per comment, and the new fallback adds another request on failure paths.

Could we make sure every response body is closed after reading it? Under CI load, leaving bodies open can cause connection/resource pressure and prevent proper connection reuse.

File: pkg/commenter/gitlab/gitlab.go [hunk 6]
```diff
@@ -126,9 +159,8 @@ func (c *Gitlab) WriteLineComment(file, comment string, line int) error {
...
+			if shouldUseGeneralPrComment(file, string(b)) {
```
Comment (mjshastha) [inline pkg/commenter/gitlab/gitlab.go]: If the inline discussion fails and the notes fallback also fails, the error appears to report the discussion response body rather than the /notes response.

Could we capture the notes response separately and return that error? Otherwise the production log can point us at the wrong API failure.

File: pkg/commenter/gitlab/gitlab.go
Comment (mjshastha) [inline pkg/commenter/gitlab/gitlab.go]: This PR increases the number of cases that can fall back to /notes, but the cleanup logic here appears to only clean /discussions.

On a subsequent scan of the same MR, could the previous fallback notes remain and create duplicates?

If so, should we extend the cleanup to the notes API, or is the persistence of these fallback comments intentional?

Comment (mjshastha) [inline pkg/commenter/gitlab/gitlab.go]: Jira explicitly mentions silencing the error when the comments to be added are null/empty, but I don't see a guard before we start making the GitLab API calls.

Can we return early for an empty/whitespace-only comment? This would avoid unnecessary /versions and discussion requests as well.

If this is expected to be handled upstream in trivy-plugin, can we point to that check so we can confirm the Jira requirement is covered?

**Review by mjshastha — COMMENTED** [review]
The failure reported in SLK-132403 happens in getLatestVersion() before we reach the new fallback logic. If GitLab returns the elevated 400/rate-limit response from the Jira example, this path will still fail exactly as before.
